_posts/2024-01-29-good-parallel-computer.md

```diff
@@ -0,0 +1,74 @@
+
+[Work graphs] also fit into the category of complexifying the execution model to work around the limitations of the primitive 3D grid. I was initially excited about their prospect, but when I took a closer look, I found they were inadequate for expressing any of the producer/consumer relationships in Vello, for three reasons: lack of joins, no ability to maintain ordering constraints, and fixed size inputs only.
+
+There's a lot of accidental complexity as well. There are multiple competing APIs, each with subtly different semantics, which makes it especially hard to write code once and just have it work.
+
+The complexity of the GPU ecosystem has many downstream effects. Drivers and shader compilers are buggy and [insecure], and there is probably no path to really fixing that. Core APIs tend to be very limited in functionality and performance, so there's a dazzling array of extensions that need to be detected at runtime, and the most appropriate permuation selected. This in turn makes it far more likely to run into bugs that appear only with specific combinations of features, or on particular hardware.
+
+All this is in fairly stark contrast to the CPU world. A modern CPU is also dazzlingly complex, with billions of transistors, but it is rooted in a much simpler computational model. From a programmer perspective, coding for an Apple M3 isn't that different than, say, a Cortex M0, which can be made with about 48,000 transistors. Similarly, a low performance RISC-V implementation is a reasonable student project. Obviously the M3 is doing a lot more with branch prediction, superscalar issue, memory hierarchies, op fusion, and other performance tricks, but it's recognizably doing the same thing as a vastly similar chip.
+
+In the past, there were economic pressures towards replacing special-purpose circuitry with general purpose compute performance, but those incentives are shifting. Basically, if you're optimizing for number of transistors, then somewhat less efficient general purpose compute can be kept busy almost all the time, while special purpose hardware is only justified if there is high enough utilization in the workload. However, as Dennard scaling has ended and we're more constrained by power than transistor count, special purpose hardware starts winning more; it can simply be powered down if it isn't used by the workload. The days of a purely RISC computational model are probably over. What I'd *like* to see replacing it is an agile core (likely RISC-V) serving as the control function for a bunch of special-purpose accelerator extensions. That certainly is the model of the [Vortex] project among others.
+
```

Comment:
As a counterpoint, in the world of Big Data, we kinda have this. You write kernel-like functions, upload them to your favorite parallel system (e.g. Snowflake) and run them on zillions of cores and I/O channels. It works because the logic is kept pretty simple, at least in the inner loop, and the universal goal is to radically reduce the size of the dataset as early as possible in the pipeline. Query optimizers hide complexity on one end, and application logic on the other.

"Agility" is compromised but that's taken as fundamental to parallel computing.